src/lib/krb5/os/dnssrv.c

```diff
@@ -45,11 +45,128 @@ krb5int_free_srv_dns_data (struct srv_dns_entry *p)
     }
 }
 
-/* Do DNS SRV query, return results in *answers.
+/*
+ * Prepares a DNS label of service.protocol.realm using a caller-supplied fixed
+ * buffer.  Since the protocol is not relevant to all callers it can be set to
+ * NULL to be excluded from the label.
+ */
+static krb5_error_code
+prepare_lookup_buf(const krb5_data *realm, const char *service,
+                   const char *protocol, char *fixed_buf, size_t bufsize)
+{
+    struct k5buf buf;
+
+    if (memchr(realm->data, 0, realm->length))
+        return EINVAL;
+
+    k5_buf_init_fixed(&buf, fixed_buf, bufsize);
+    k5_buf_add_fmt(&buf, "%s.", service);
+    if (protocol != NULL)
+        k5_buf_add_fmt(&buf, "%s.", protocol);
+    k5_buf_add_len(&buf, realm->data, realm->length);
+
+    /*
+     * Realm names don't (normally) end with ".", but if the query doesn't end
+     * with "." and doesn't get an answer as is, the resolv code will try
+     * appending the local domain.  Since the realm names are absolutes, let's
+     * stop that.
+     */
+
+    if (buf.len > 0 && ((char *)buf.data)[buf.len - 1] != '.')
+        k5_buf_add(&buf, ".");
+
+    return k5_buf_status(&buf);
+}
+
+/* Insert into the list, ordering by priority.  Weight is not currently
+ * used. */
```

Comment:
This should be more specific about the parameters: "Insert new into the list *head, ordering..."